lib/dal/src/action/dependency_graph.rs

```diff
@@ -0,0 +1,122 @@
+use std::collections::{HashMap, HashSet};
+
+use crate::{
+    action::Action, dependency_graph::DependencyGraph, ActionId, Component, ComponentId, DalContext,
+};
+
+use super::ActionResult;
+
+#[derive(Debug, Clone)]
+pub struct ActionDependencyGraph {
+    inner: DependencyGraph<ActionId>,
+}
+
+impl Default for ActionDependencyGraph {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ActionDependencyGraph {
+    pub fn new() -> Self {
+        Self {
+            inner: DependencyGraph::new(),
+        }
+    }
+
+    /// Construct an [`ActionDependencyGraph`] of all of the queued [`Action`s][crate::action::Action]
+    /// for the current [`WorkspaceSnapshot`][crate::WorkspaceSnapshot].
+    pub async fn for_workspace(ctx: &DalContext) -> ActionResult<Self> {
+        // * Get all ActionId -> ComponentId mappings.
+        // * For each of these ComponentIds (A):
+        //     * For each Input Socket:
+        //         * For each source ComponentId (B):
+        //           * All Actions for Component A depend on All actions for Component B
+        let mut component_dependencies: HashMap<ComponentId, HashSet<ComponentId>> = HashMap::new();
+        let mut actions_by_component_id: HashMap<ComponentId, HashSet<ActionId>> = HashMap::new();
+        let mut action_dependency_graph = Self::new();
+
+        // Need to get all actions that are still in the "queue", including those that have failed,
+        // or are currently running.
+        for action_id in Action::all_ids(ctx).await? {
+            action_dependency_graph.inner.add_id(action_id);
+
+            // Theoretically, we may have Actions at some point that aren't Component specific.
+            if let Some(component_id) = Action::component_id(ctx, action_id).await? {
+                actions_by_component_id
+                    .entry(component_id)
+                    .or_default()
+                    .insert(action_id);
+            }
+        }
+
+        // TODO: Account for explicitly defiend dependencies between actions. These should be edges
+        //       directly between two Actions, but are not implemented yet.
+
+        // Action dependencies are primarily based on the data flow between Components. Things that
+        // feed data into other things must have their actions run before the actions for the
+        // things they are feeding data into.
+        for component_id in actions_by_component_id.keys().copied() {
+            let component = Component::get_by_id(ctx, component_id).await?;
+            for incoming_connection in component.incoming_connections(ctx).await? {
+                component_dependencies
+                    .entry(component_id)
+                    .or_default()
+                    .insert(incoming_connection.from_component_id);
+            }
+            // TODO: Handle implicit/inferred dependencies from Frames once that work lands?
```

Comment:
You should be able to do the same as above, only via `component.inferrred_connections` but you'll need to filter for deletes as I couldn't make this check too low level without hiding the `to_delete` flag on the incoming inferred connection

Reply:
`component.inferred_connections` has both the incoming, and outgoing inferred connections, right? So, it would need to be limited to inferred connections where the `to_component_id` is the same as `component.id()`. Looks easy enough, though. 👍🏻